### PR TITLE
Fix: multiple memory leak and array access out of bound

### PR DESCRIPTION
**Pull request by caic99**
### Reminder
- [ ] Have you linked an issue with this pull request?
- [ ] Have you added adequate unit tests and/or case tests for your pull request?
- [ ] Have you noticed possible changes of behavior below or in the linked issue?
- [ ] Have you explained the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi? (ignore if not applicable)

### Linked Issue
#3553

### CODE REVIEW

File: source/module_basis/module_nao/projgen.cpp [hunk 1]
```diff
@@ -23,7 +23,7 @@ void projgen(const int l,
     assert(rcut < r[nr - 1]);
     assert(std::is_sorted(r, r + nr));
 
-    std::vector<double> dr(nr - 1);
+    std::vector<double> dr(nr);
```
Comment (mohanchen) [inline source/module_basis/module_nao/projgen.cpp:26]: why is this 'nr'?

File: source/module_esolver/esolver_ks_lcao_tddft.cpp [hunk 2]
```diff
@@ -48,17 +48,17 @@ ESolver_KS_LCAO_TDDFT::~ESolver_KS_LCAO_TDDFT()
     {
         for (int ik = 0; ik < kv.nks; ++ik)
         {
-            delete Hk_laststep[ik];
+            delete[] Hk_laststep[ik];
```
Comment (mohanchen) [inline source/module_esolver/esolver_ks_lcao_tddft.cpp:51]: Excellent! I am curious how did you catch this bug?

Comment (caic99) [reply source/module_esolver/esolver_ks_lcao_tddft.cpp:51]: > Excellent! I am curious how did you catch this bug?

These bugs are shown in the [dynamic analysis report](https://deepmodeling.github.io/abacus-develop/).

File: source/module_basis/module_nao/projgen.cpp (hunk 1 above)
Comment (caic99) [reply source/module_basis/module_nao/projgen.cpp:26]: > why is this 'nr'?

`std::adjacent_difference` returns with the same size as input.

**Review by WHUweiqingzhou — APPROVED** [review]
Good job! LGTM.

**Pull request merged**